### PR TITLE
Add storage pool state and service packages for managing storage pools.

### PR DESCRIPTION
**Pull request by wallyworld**
This PR adds the domain and service packages for managing storage pools. The main use cases to support  are CRUD from the CLI (including querying based on a names and types filter), as well as getting a pool by name for use in the provisioning and deploy facades.

The service needs a registry representing the valid storage provisioners for the current model - this is passed in when the service is constructed.

## QA steps

just unit tests at this stage

## Links

**Jira card:** JUJU-5579

### CODE REVIEW

File: domain/storage/state/state.go [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Trace(domain.CoerceError(err))
+		}
+		if err == nil {
+			return fmt.Errorf("storage pool %q %w", pool.Name, errors.AlreadyExists)
+		}
+		poolUUID := uuid.MustNewUUID().String()
+
+		if err := upsertStoragePool(ctx, tx, poolUUID, pool.Name, pool.Provider); err != nil {
+			return errors.Annotate(domain.CoerceError(err), "creating storage pool")
```
Comment (tlm) [inline domain/storage/state/state.go:58]: Can we include some context in the error message. i.e name

File: domain/storage/state/state.go [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+		}
+
+		if err := updatePoolAttributes(ctx, tx, poolUUID, pool.Attrs); err != nil {
+			return errors.Annotatef(domain.CoerceError(err), "creating storage pool %s attributes", poolUUID)
```
Comment (tlm) [inline domain/storage/state/state.go:62]: Can we include name and get rid of UUID? UUID is made inside of the function and unless this function returns successfully it has zero meaning.

File: domain/storage/state/state.go [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+		}
+		return nil
+	})
+
+	return errors.Trace(err)
```
Comment (tlm) [inline domain/storage/state/state.go:67]: I think CoerceError here and get rid of all the ones above. Just my two cents though.

File: domain/storage/state/state.go [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+}
+
+func upsertStoragePool(ctx context.Context, tx *sqlair.TX, poolUUID string, name string, providerType string) error {
```
Comment (tlm) [inline domain/storage/state/state.go:70]: Can you make this into a closure return that then takes the tx. We don't want to be doing prepares in the tx.

File: domain/storage/state/state.go [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+	}
+
+	err = tx.Query(ctx, insertStmt, dbPool).Run()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func updatePoolAttributes(ctx context.Context, tx *sqlair.TX, storagePoolUUID string, attr domainstorage.Attrs) error {
```
Comment (tlm) [inline domain/storage/state/state.go:107]: Same comment above. Need to return a closure.

File: domain/storage/state/state.go [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+		return errors.Trace(err)
+	}
+
+	for key, value := range attr {
+		if err := tx.Query(ctx, insertStmt, poolAttribute{
+			ID:    storagePoolUUID,
+			Key:   key,
+			Value: value,
+		}).Run(); err != nil {
+			return errors.Trace(domain.CoerceError(err))
```
Comment (tlm) [inline domain/storage/state/state.go:145]: I think it should be the callers responsibility to Coerce as we are still in the state layer. More two cents.

File: domain/storage/state/state.go [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+		if err := updatePoolAttributes(ctx, tx, poolUUID, pool.Attrs); err != nil {
+			return errors.Annotatef(domain.CoerceError(err), "updating storage pool %s attributes", poolUUID)
+		}
+		return nil
+	})
+
+	return errors.Trace(err)
+}
+
+func (st State) loadStoragePools(ctx context.Context, tx *sqlair.TX, filter domainstorage.StoragePoolFilter) ([]domainstorage.StoragePoolDetails, error) {
```
Comment (tlm) [inline domain/storage/state/state.go:234]: Another closure please.

File: domain/storage/state/state.go [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+		storagePools, err = st.loadStoragePools(ctx, tx, domainstorage.StoragePoolFilter{
+			Names: []string{name},
+		})
+		return errors.Trace(err)
+	})
+	if err != nil {
+		return domainstorage.StoragePoolDetails{}, errors.Trace(err)
+	}
+	if len(storagePools) == 0 {
+		return domainstorage.StoragePoolDetails{}, fmt.Errorf("storage pool %q %w", name, errors.NotFound)
```
Comment (tlm) [inline domain/storage/state/state.go:326]: Domain specific error would make this duck very happy.

File: domain/storage/service/service.go [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+func NewService(st State, logger Logger, registry storage.ProviderRegistry) *Service {
+	return &Service{
+		st:       st,
+		logger:   logger,
+		registry: registry,
+	}
+}
+
+// PoolAttrs define the attributes of a storage pool.
+type PoolAttrs map[string]any
```
Comment (tlm) [inline domain/storage/service/service.go:49]: You know this code better than me. But should we rip the band aid off now and just make this [string]string and get some type safety back in our lives?

It is up to the provider validation logic to infer meaning.

**Review by tlm — APPROVED** [review]
Thanks @wallyworld this is looking good. Happy to chat if you want.

File: domain/storage/service/service.go (hunk 9 above)
Comment (wallyworld) [reply domain/storage/service/service.go:49]: the args come over the wire as any (as supplied by the cli and client callers). the facade just hands the args as received to the service. the service is responsible for validation and coercion.

File: domain/storage/service/service.go [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+
+// CreateStoragePool creates a storage pool, returning an error satisfying [errors.AlreadyExists]
+// if a pool with the same name already exists.
+func (s *Service) CreateStoragePool(ctx context.Context, name string, providerType storage.ProviderType, attrs PoolAttrs) error {
+	if name == "" {
+		return storageerrors.MissingPoolNameError
+	}
+	if providerType == "" {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/storage/service/service.go:57]: No, it's just a type alias. Its value is very cloud specific. It's the same reason why it's just a string in the table (see schema comment)

File: domain/storage/service/service.go (hunk 9 above)
Comment (wallyworld) [reply domain/storage/service/service.go:49]: the facade could coerce to strings off the wire.

File: domain/storage/service/service.go [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+	sp := domainstorage.StoragePoolDetails{
+		Name:     name,
+		Provider: string(providerType),
+		Attrs:    attrsToSave,
+	}
+	err = s.st.CreateStoragePool(ctx, sp)
+	return errors.Annotatef(err, "creating storage pool %q", name)
+}
+
+func (s *Service) validateConfig(name string, providerType storage.ProviderType, attrs map[string]interface{}) error {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/storage/service/service.go:76]: Can do, this was transplanted business logic with legacy error handling, but a typed error is good.

File: domain/storage/service/service_test.go [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+
+	err := s.service().CreateStoragePool(context.Background(), "ebs-fast", "ebs", PoolAttrs{"foo": "foo val"})
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *serviceSuite) TestCreateStoragePoolValidates(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	err := s.service().CreateStoragePool(context.Background(), "ebs-fast", "ebs", PoolAttrs{"bar": "bar val"})
+	c.Assert(err, gc.ErrorMatches, `.* missing attribute foo`)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/storage/service/service_test.go:100]: This is an error generated by the test, not business logic. I can update the test to add a typed error but it's not the point of the test here.

File: domain/storage/service/service.go (hunk 9 above)
Comment (tlm) [reply domain/storage/service/service.go:49]: no worries. Thank you for answering. I love it when the client send `any` over the wire. What could go wrong.

File: domain/storage/types.go [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+// StoragePoolDetails defines the details of a storage pool to save.
+// This type is also used when returning query results from state.
+type StoragePoolDetails struct {
+	Name     string
+	Provider string
+	Attrs    Attrs
+}
+
+// StoragePoolFilter defines attributes used to filter storage pools.
+type StoragePoolFilter struct {
```
Comment (manadart) [inline domain/storage/types.go:18]: What are the use-cases for this?

**Pull request merged**